Max torch final victory dialog handling of SoF s9 rework

max-torch opened the pull request:
This is my contribution to completing what was started in #6334 .
Things in the scenario didn't quite make logical sense considering the presence of Gryphon Riders. This resolves the issue by allowing them to trigger an event, and by acknowledging them in the `endlevel` dialog.

stevecotton commented on data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg:
```diff
@@ -672,15 +671,227 @@
+                [else]
+                    [message]
+                        speaker=$unit.id
+                        message= _ "The gryphon rider puzzled over a stone tile sitting loose on the small mesa. Beneath a spattering of bat guano, there was a glowing rune. The rider didn’t know what to make of it, so the rider left it for now."
```
The grammar feels a bit odd, as if "rider" refers to the gryphon.

"Neither mount nor rider knew what to make of it, so they left it for now."

stevecotton commented on data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg:
```diff
@@ -672,15 +671,227 @@
             [/message]
+            [message]
+                speaker=$unit.id
+                message= _ "Alright, there's no turning back now!"
```
That's a non-typographic apostrophe.

Wedge009 commented on data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg:
```diff
@@ -672,15 +671,227 @@
             [message]
                 speaker=narrator
-                message= _ "Krawg puzzled over a stone tile sitting loose on the small mesa. Beneath a spattering of bat guano, there was a glowing rune. Krawg didn’t know what to make of it, so the gryphon left it for now."
+                message= _ "Once the magic seal capping the throat of the volcano was removed, lava and toxic gas exploded upward, killing everyone inside and sealing them in a tomb of hot, flowing rock. All of the gryphon riders flew up towards a hole in the roof with utmost haste, but even they could not outrun the eruption. Only Krawg managed to barely escape due to his speed and the absence o ...
```
> Only Krawg managed to barely escape due to his speed and the absence of the weight of a rider on his back.

Not a big issue, but I feel this is unnecessarily explicit in its explanation. Maybe it's sufficient just to say Krawg was the only one fast enough to escape and let the player infer the rest.

(This string is repeated on line 1088.)

Wedge009 commented:
I tested this and it seems to cover the issues we discussed previously: a non-Krawg unit finding the final rune and also addresses the plot-hole regarding the Gryphon Riders. A bit unfortunate that the entire volcano eruption event has to be duplicated to cover both the Krawg version and the regular Gryphon Rider version - might it be possible to re-use that code?

At any rate, as it currently stands it seems to resolve #6332 satisfactorily.

max-torch commented:
>might it be possible to re-use that code?

Thank you for testing and confirming on your end. I'll see about doing one more refinement of the dialog. As to minimizing duplication of code, I'll see what I can do. Perhaps define the code as a function/macro and then just call it for both conditions.

stevecotton commented on data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg:
```diff
@@ -672,15 +671,227 @@
+            [/remove_item]
+            [floating_text]
+                x,y=29,31
+                text="<span color='#ff7722'>" + _ "Smash!" + "</span>"
```
Assuming I've understood the meaning correctly, `# po: Treat this as if it was a sound-effect in a comic book`

knyghtmare commented on data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg:
```diff
@@ -867,11 +1078,25 @@
-            [/message]
+            [if]
+                [have_unit]
+                    type=Gryphon Rider
```
Should be
```ini
type=Gryphon Rider,Gryphon Master
```

Wedge009 replied:
Oh, I missed that - and I tested with a Gryphon Master...

knyghtmare commented:
also, be sure to run `wmlindent` on this to get all 12 checks on the pesky CI.

max-torch commented:
I would like to provide some replays to show how I have tested the latest commit.
**Case 1: No other gryphons in the level, and Krawg gets tablet**
[SoF-Caverns of Flame replay 20220421-180422.gz](https://github.com/wesnoth/wesnoth/files/8530119/SoF-Caverns.of.Flame.replay.20220421-180422.gz)
**Case 2: Level ends with Gryphon Master at the center, and Krawg gets tablet**
[SoF-Caverns of Flame replay 20220421-180536.gz](https://github.com/wesnoth/wesnoth/files/8530135/SoF-Caverns.of.Flame.replay.20220421-180536.gz)
**Case 3: No gryphons at the center, and a Gryphon Master gets the tablet**
[SoF-Caverns of Flame replay 20220421-180700.gz](https://github.com/wesnoth/wesnoth/files/8530148/SoF-Caverns.of.Flame.replay.20220421-180700.gz)
**Case 4: A gryphon is at the center, and a Gryphon Master gets the tablet**
[SoF-Caverns of Flame replay 20220421-180802.gz](https://github.com/wesnoth/wesnoth/files/8530163/SoF-Caverns.of.Flame.replay.20220421-180802.gz)

max-torch commented:
> Also, I forgot to mention: would it be reasonable to use 'out-fly' instead of 'outrun'?

After a quick internet search, outrun seems to be used in a lot of contexts that don't involve literal running. Seems like it can be used in the context of anything that travels faster. Merriam-Webster lists outfly as a word but the Cambridge Dictionary doesn't. I found a graph showing outfly had little usage over time. So, for the sake of being less obscure, I guess I'm in favor of outrun.

max-torch commented:
The PR is now squashed neatly into a single commit and rebased on top of the HEAD of the present Master branch. After this passes the CI (hopefully optimistic) I submit this for merging.

Wedge009 commented:
Looks good to me. Thanks.

knyghtmare reviewed (approved):
I check code but I feel like Krawg has way too much plot armour here, but I guess it's fine.

Wedge009 commented:
The plot armour was always there - at least since the scenario was rewritten in 1.15.x - my request and this PR are trying to address the game-play issue of Krawg being the only one permitted to break the last rune and the story issue of the Gryphon Riders not being able to escape.

knyghtmare commented:
> The plot armour was always there - at least since the scenario was rewritten in 1.15.x - my request and this PR are trying to address the game-play issue of Krawg being the only one permitted to break the last rune and the story issue of the Gryphon Riders not being able to escape.

One thing I like about this reworked SoF is that it removed that non-sensical concept of "Thursagan dying and then all other runecrafters dying with him"..

But I dont like SoF's high difficulty curve (since I cannot get pass the 5th level despite 20ish tries)

Wedge009 commented:
Thanks everyone for your contributions.

stevecotton commented on data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg:
```diff
@@ -701,49 +714,100 @@
+                    [/message]
+                    [message]
+                        speaker=Thursagan
+                        message= _ "Durstorn and Glildur have both convinced me that the time is not right for the Sceptre of Fire, best that it rests with us for a while. Now, smash it!"
```
Late comment, because I only realised when reviewing the .po file fuzzies:
```
# po: The "smash it" refers to a stone tablet, not the Sceptre.
```

max-torch replied:
Thanks for handling this in #6633.

stevecotton commented on data/campaigns/Sceptre_of_Fire/scenarios/9_Caverns_of_Flame.cfg:
```diff
@@ -701,49 +714,100 @@
+                    [/message]
+                    [message]
+                        speaker=Thursagan
+                        message= _ "Just do it! You riders have a chance to escape, our enemies cannot. We knew that death was likely when we fled here, just go!"
```
This is one of the ones that might have another variant if there's only one gryphon rider, maybe: "Just do it! You and Krawg have a change to escape, our enemies cannot. We knew that death was likely when we fled here, just go!".
